Bug 1771809: Add e2e for PodDisruptionBudgetAtLimit alert

sallyom opened the pull request:
Here's an e2e that:
* creates a pod
* creates a PodDisruptionBudget  of minAvail=1
* queries prometheus for a pending 'PodDisruptionBudgetAtLimit' alert 'pending' - to wait for 'firing' you'd have to wait 15 min, so pending it is. 
* query uses helper functions from this PR: https://github.com/openshift/library-go/pull/602 (fake bump here) 

/cc @soltysh @marun

sallyom commented:
/test e2e-aws-operator

sanchezl commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+	kcmoSigningKeySecret = "kube-controller-manager-operator-serving-cert"
+)
+
+func TestE2E(t *testing.T) {
```
I'm not sure this nesting is needed. All tests under `test/e2e` are e2e tests, and for example , `TestSATokenSignerControllerSyncCerts` is in another file and won't be in this hierarchy.

marun replied:
Nesting tests allows setup in the enclosing function to be shared, but I would agree that unless the setup is expensive (e.g. creating resources) I'm not sure it makes sense to use nesting vs creating a helper that tests can call.

marun replied:
@sanchezl Ah, and to your point, naming this `TestE2E` probably doesn't make sense unless this is the entrypoint for running all e2e tests rather than just some of them.

sallyom replied:
updated, took out the nesting.

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
 	}
+
+	// test that operator namespace exists
+	t.Run("OperatorNamespace", func(t *testing.T) {
```
Is the operator namespace being present a precondition for the other test to execute? Since `t.Run` isolates failure, enclosing this check would result in suite failure but would not prevent `PodDisruptionBudgetAtLimit alert` from executing.

sallyom replied:
Good point, I removed the nesting.

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+	})
+
+	// test that pdb-atlimit alert exists when there is a pdb at limit
+	t.Run("PodDisruptionBudgetAtLimit alert", func(t *testing.T) {
```
(No action required) Consider using of dashes or underscores in test names instead of spaces to simplify targeting test execution.

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+	client.CoreV1().Namespaces().Delete(ns, &metav1.DeleteOptions{})
+}
+
+// took this from service-ca-operator test
```
Suggesting using uuid instead:

https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/vendor/github.com/openshift/library-go/pkg/config/leaderelection/leaderelection.go#L32

marun replied:
But this isn't really the best way to deal with ensuring name uniqueness in tests. Setting `ObjectMeta.GenerateName` is likely preferable if it's possible to create the object before knowing the name:

https://github.com/kubernetes-sigs/kubefed/blob/master/test/e2e/framework/framework.go#L308

It's also possible to directly call the apiserver's name generating logic:

https://github.com/kubernetes-sigs/kubefed/blob/master/test/e2e/defaulting.go#L66

sallyom replied:
looking into it...

sallyom replied:
ah, I knew there must be a better way.  fixed.

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+func getBearerTokenURLViaPod(t *testing.T, kclient *kubernetes.Clientset, ns, execPodName, url, bearer string) ([]byte, error) {
+	//cmd := fmt.Sprintf("curl -s -k -H 'Authorization: Bearer %s' %q", bearer, url)
+	// Retrieve the cert and key PEM of the current CA via kube-controller-manager-operator serving-cert secret
+	secret, err := kclient.CoreV1().Secrets(kcmoNamespace).Get(kcmoSigningKeySecret, metav1.GetOptions{})
```
That configuring tls with the serving cert works is an accident of implementation. The signing CA is included with the serving cert, allowing it to be reused as a ca bundle. However, this is not good practice and at some point in the future serving certs will not include the signing CA. Please source an actual ca bundle from the service ca operator instead as per the following example (create annotated configmap, wait ca bundle to be added to configmap):

https://github.com/openshift/service-ca-operator/blob/ca-validity-duration/test/e2e/e2e_test.go#L329

sallyom replied:
ok, got it.. working on it...

sallyom replied:
done, i'm obtaining the caBundle

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+		return "", "", false
+	}
+
+	for i := 0; i < 30; i++ {
```
Please use one of the [wait.Poll](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go#L333) functions instead of a manual loop e.g.

https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/vendor/github.com/openshift/library-go/pkg/operator/certrotation/config.go#L17

sallyom replied:
will do

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+	})
+}
+
+func pdbCreate(t *testing.T, client *policyclientv1beta1.PolicyV1beta1Client, name string, labels map[string]string) error {
```
`t` is supplied to this method and the one that follows but does not appear to be used (maybe a linter is in order).

sallyom replied:
yes, correct, cleaning that up

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{Name: "test", Image: "centos:7", Command: []string{"sh", "-c", "trap exit TERM; while true; do sleep 5; done"}},
```
Please prefer breaking long lines on parameters for readability, e.g

```golang
{
  Name: "test", 
  Image: "centos:7", 
  Command: []string{
     "sh", 
     "-c", 
     "trap exit TERM; while true; do sleep 5; done",
  },
}
```

sallyom replied:
ok, done

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+			t.Fatal(err)
+		}
+
+		fmt.Println("Waiting 15s for PodDisruptionBudget to process")
```
Why is it desirable to sleep for 15 seconds vs just relying on the subsequent polling loop?

sallyom replied:
I've added a wait.Poll for this

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+		}
+		var pods *corev1.PodList
+		// Poll to see if pod is created
+		wait.Poll(time.Second*5, time.Second*300, func() (bool, error) {
```
If line 71 executed successfully, what would cause the pod to disappear?

sallyom replied:
wait.Poll for "Running" (I added that), not sure it's necessary, though

marun commented on test/e2e/operator_test.go:
```diff
@@ -1,26 +1,316 @@
+func runQuery(t *testing.T, kclient *kubernetes.Clientset, query, ns, execPodName, baseURL, bearerToken string, expected bool) error {
+	// expect all correct metrics within a reasonable time period
+	var err error
+	for i := 0; i < waitForPrometheusStartSeconds; i++ {
```
Please prefer a `wait.Poll` method.

sallyom replied:
added wait.Poll there

sallyom commented:
@marun @sanchezl thanks for the review! I've addressed your suggestions, let me know if this (or more generic obtain caBundle/query functions, at least) should be moved to library-go.

sallyom commented:
/test e2e-aws

sttts commented on go.mod:
```diff
@@ -36,7 +37,7 @@ replace (
 	github.com/openshift/api => github.com/openshift/api v0.0.0-20191101174058-d428681bd3ce
 	github.com/openshift/client-go => github.com/openshift/client-go v0.0.0-20191022152013-2823239d2298
-	github.com/openshift/library-go => github.com/openshift/library-go v0.0.0-20191121124438-7c776f7cc17a
+	github.com/openshift/library-go => github.com/sallyom/library-go v0.0.0-20191125154451-7578d03d01df
```
only for testing?

sallyom replied:
yes, only for testing, will remove once library-go pr is merged.. i'll hold until then.

sttts commented on test/e2e/operator_test.go:
```diff
@@ -24,3 +35,129 @@ func TestOperatorNamespace(t *testing.T) {
+		t.Fatal(err)
+	}
+	// Now check for alert
+	wait.Poll(time.Second*5, time.Second*300, func() (bool, error) {
```
wait.InfiniteTestTimeout or something like that instead of 5min

sttts replied:
same above

sallyom replied:
I've moved to `wait.PollImmediate(time.Second*1, testTimeout, func() (bool, error) {`
and testTimeout 120s.  Test time was 25s (oops didn't realize how long that was) down to 14s.

sttts commented on test/e2e/operator_test.go:
```diff
@@ -24,3 +35,129 @@ func TestOperatorNamespace(t *testing.T) {
+			return false, err
+		}
+		if len(pods.Items) > 0 {
+			if pods.Items[0].Status.Phase == corev1.PodRunning {
```
combine into one if

sallyom replied:
done

sttts commented on test/e2e/operator_test.go:
```diff
@@ -24,3 +35,129 @@ func TestOperatorNamespace(t *testing.T) {
+
+// TestPodDisruptionBudgetAtLimitAlert tests that pdb-atlimit alert exists when there is a pdb at limit
+func TestPodDisruptionBudgetAtLimitAlert(t *testing.T) {
+	// See https://bugzilla.redhat.com/show_bug.cgi?id=1762888
```
nit: move into func comment

sallyom replied:
done

sttts commented on test/e2e/operator_test.go:
```diff
@@ -1,15 +1,26 @@
+	metrics "github.com/openshift/library-go/pkg/metrics"
 )
 
+const pdbAlertQuery = `ALERTS{alertstate="pending",alertname="PodDisruptionBudgetAtLimit",severity="warning"} == 1`
```
would inline this into the func.

sallyom replied:
ok, done

sallyom commented:
/hold 

until fake bump is removed, library-go pr merged: https://github.com/openshift/library-go/pull/602

sallyom commented:
@sttts thanks for the review, i've updated, ready for another look.

sallyom commented:
/bugzilla refresh

sallyom commented:
/bugzilla refresh

marun commented on test/e2e/operator_test.go:
```diff
@@ -24,3 +33,136 @@ func TestOperatorNamespace(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	caBundle, err := metrics.RetrieveServiceCABundle(kubeClient, cmName, ns.Name)
```
I'm happy to see this decomposed implementation - I think it will prove easier to maintain. I think the next step is adding a helper method to simplify calling for cases where only a single query is required e.g.

```golang
metrics.QueryPrometheus(kubeClient, ns.Name, query)
```

marun commented on test/e2e/operator_test.go:
```diff
@@ -24,3 +33,136 @@ func TestOperatorNamespace(t *testing.T) {
+		t.Fatal(err)
+	}
+	// Now check for alert
+	wait.PollImmediate(time.Second*1, testTimeout, func() (bool, error) {
```
Is this polling loop necessary given that `RunPrometheusQuery` is already running a polling loop?

sallyom replied:
nope, updated : )

sallyom commented:
I notice the CI template cluster-launch-installer-src test pod does not have access to oc/kubectl, so will have to work around that or add the cli to the CI template. (that's why the aws-operator test fails here, tests run locally fine as long as oc/kubectl is on the system)

marun commented on test/e2e/operator_test.go:
```diff
@@ -24,3 +38,126 @@ func TestOperatorNamespace(t *testing.T) {
+	wait.PollImmediate(time.Second*1, testTimeout, func() (bool, error) {
+		pods, err = kubeClient.CoreV1().Pods(name).List(metav1.ListOptions{LabelSelector: "app=pdbtest"})
+		if err != nil {
+			return false, err
```
Is this error fatal?

sallyom commented:
/test e2e-aws

sallyom commented:
/test e2e-aws

sallyom commented:
/bugzilla refresh

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
/test e2e-aws

soltysh commented on Makefile:
```diff
@@ -34,6 +34,7 @@ $(call build-image,ocp-cluster-kube-controller-manager-operator,$(IMAGE_REGISTRY
 $(call add-bindata,v4.1.0,./bindata/v4.1.0/...,bindata,v411_00_assets,pkg/operator/v411_00_assets/bindata.go)
 
 test-e2e: GO_TEST_PACKAGES :=./test/e2e/...
+test-e2e: GO_TEST_FLAGS += -v
```
Ditto as in the other PR, I don't think we need it here.

sallyom replied:
I prefer to see what's going on rather than a hang, but I removed this change : )

soltysh commented on test/e2e/operator_test.go:
```diff
@@ -24,3 +33,136 @@ func TestOperatorNamespace(t *testing.T) {
+	// Wait for pbd
+	err = wait.PollImmediate(time.Second*1, testTimeout, func() (bool, error) {
+		if _, err := policyClient.PodDisruptionBudgets(name).List(metav1.ListOptions{LabelSelector: "app=pbtest"}); err != nil {
+			return false, fmt.Errorf("Waiting for poddisruptionbudget")
```
Now with go1.13:
```
fmt.Errorf("waiting for poddisruptionbudget: %w", err)
```

soltysh commented on test/e2e/operator_test.go:
```diff
@@ -24,3 +33,136 @@ func TestOperatorNamespace(t *testing.T) {
+		if err != nil {
+			return false, err
+		}
+		if len(pods.Items) > 0 && pods.Items[0].Status.Phase == corev1.PodRunning {
```
This still holds.

soltysh commented on test/e2e/operator_test.go:
```diff
@@ -61,3 +164,46 @@ func TestKCMRecovery(t *testing.T) {
+		},
+	}
+	_, err := client.CoreV1().Pods(name).Create(pod)
+	if err != nil {
```
Just `return err`

sallyom replied:
done

soltysh commented:
/hold cancel

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
/retest

soltysh commented:
/retest
